ESPS/general/src/fea2esig/fea2esig.1
.\" Copyright (c) 1995 Entropic Research Laboratory, Inc. All rights reserved.
.\" @(#)fea2esig.1	1.8 27 Sep 1997 ERL
.ds ]W (c) 1993 Entropic Research Laboratory, Inc.
.if t .ds - \-\h'-0.2m'\-
.if n .ds - ---
.TH FEA2ESIG 1\-ESPS 27 Sep 1997
.SH NAME
.nf
fea2esig \- write FEA file data to Esignal file
.fi
.SH SYNOPSIS
.B fea2esig
[
.BI \-a " format"
] [
.BI \-f " field"
] .\ .\ . [
.BI \-r " range"
] [
.BI \-x " debug_level"
] [
.BI \-A " width"
] [
.BI \-F
]
.I input.fea output.esig
.SH DESCRIPTION
.PP
.I fea2esig
reads a file
.I input.fea
in ESPS FEA format (see
.IR ESPS (5\-ESPS),
.IR FEA (5\-ESPS))
and writes an output file
.I output.esig
in
.I Esignal
format (see Ref. [1]) containing the same information
(or a subset thereof).
If
.I input.fea
is "\-", standard input is used for the input file.
If
.I output.esig
is "\-", standard output is used for the output file.
The input and output should not be the same file;
however, you can run the program as a filter
by specifying "\-" for both input and output.
.PP
Normally, the contents of all fields in all input records
are copied to the output;
one output record is written for each input record.
However, with the
.B \-r
option, one can restrict processing to a subrange of the input records,
and with the
.B \-f
option, one can restrict processing to a specified subset of the input fields.
.PP
Each ESPS FEA field in the input (if retained)
becomes an Esignal REQUIRED field with the same name in the output.
The rank and dimensions of the field are preserved.
The data type of the output is the Esignal type corresponding to
the ESPS type of the input field, as shown in the following table.
.nf
.RS
.ta 0.1i 1.3i 2.6i 4.0i

.if n \l'2.2i\-'\h'0.3i'\l'3.0i\-'
.if t \v'-0.2'\l'2.2i'\h'0.3i'\l'3.0i'\v'0.2'
	ESPS	Esignal	Esignal   	ESPS
.if n \l'2.2i\-'\h'0.3i'\l'3.0i\-'
.if t \v'-0.2'\l'2.2i'\h'0.3i'\l'3.0i'\v'0.2'
	DOUBLE	DOUBLE	DOUBLE_CPLX	DOUBLE_COMPLEX
	FLOAT	FLOAT	FLOAT_CPLX	FLOAT_COMPLEX
	LONG	LONG	LONG_CPLX	LONG_COMPLEX
	SHORT	SHORT	SHORT_CPLX	SHORT_COMPLEX
	BYTE	SCHAR	SCHAR_CPLX	BYTE_COMPLEX
	CHAR	CHAR
	CODED	SHORT
.if n \l'2.2i\-'\h'0.3i'\l'3.0i\-'
.if t \v'-0.2'\l'2.2i'\h'0.3i'\l'3.0i'\v'0.2'

.DT
.RE
.fi
The output field for an ESPS CODED field (see
.IR FEA (5\-ESPS))
has a GLOBAL subfield named "enumStrings" that is a character matrix
(type CHAR, rank 2) containing the string values of the coded type,
one per row, padded to a common length
with null characters '\\0' on the right.
Row 0 (the first row) contains the string for code 0,
row 1 contains the string for code 1, etc.
.PP
If the input file is tagged, the tags map into a REQUIRED field
named "Tag" in the output;
the field is scalar (rank 0) with data type LONG.
.PP
Each ESPS generic header item in the input
becomes an Esignal GLOBAL field with the same name in the output.
The rank of the Esignal field is 1.
Its data type is given by the following table.
.nf
.RS
.ta 0.1i 1.3i 2.6i 3.8i

.if n \l'2.2i\-'\h'0.3i'\l'2.2i\-'
.if t \v'-0.2'\l'2.2i'\h'0.3i'\l'2.2i'\v'0.2'
	Esignal	ESPS	Esignal	ESPS
.if n \l'2.2i\-'\h'0.3i'\l'2.2i\-'
.if t \v'-0.2'\l'2.2i'\h'0.3i'\l'2.2i'\v'0.2'
	DOUBLE	DOUBLE	CHAR	CHAR
	FLOAT	FLOAT	EFILE	CHAR
	LONG	LONG	AFILE	CHAR
	SHORT	SHORT	CODED	SHORT
.if n \l'2.2i\-'\h'0.3i'\l'2.2i\-'
.if t \v'-0.2'\l'2.2i'\h'0.3i'\l'2.2i'\v'0.2'

.DT
.RE
.fi
The output field for a CODED header item, like that for a CODED field,
has a GLOBAL subfield named "enumStrings" that is a character matrix
containing the string values of the coded type.
.PP
The command line with which
.I fea2esig
is invoked is converted to a character string and
added to the output file header as a GLOBAL field named "commandLine".
.PP
If the input file belongs to a special FEA subtype such as FEA_SD,
FEA_ANA, or FEA_SPEC, the output file had a scalar GLOBAL field
named "FeaSubtype" that has type SHORT and contains the subtype code
(e.g. 8 for FEA_SD, 2 for FEA_ANA, 7 for FEA_SPEC;
the complete set of codes is defined in the ESPS header file
.IR esps/fea.h ).
If defined,
.I FeaSubtype
has a GLOBAL subfield
.I enumStrings
like that for CODED FEA fields and header items.
Each row of the character matrix contains
the name of the corresponding subtype\*-e.g. "FEA_SD" in row 8,
"FEA_ANA" in row 2, etc.
Row 0 contains the string "NONE" as a placeholder, though
.I fea2esig
does not create the field
.I FeaSubtype
if the subtype code is 0.
The field
.I FeaSubtype
is created only to preserve information in the FEA header
so that the program
.IR esig2fea (1\-ESPS)
can restore the subtype when converting the Esignal file
back to a FEA file;
the notion of a FEA subtype has no exact parallel in the Esignal framework.
.PP
The output file header has a single VIRTUAL field named "source_0",
whose subfield list consists of the input header,
converted from FEA format to Esignal.
The field contains a string that gives the name of the input file,
or "<stdin>" if "\-" was specified for the input.
The source-file field list
is usually similar to the output-file field list,
but there are a number of differences.
The source-file field list records all the fields of the input file,
even if the
.B \-f
option is used to specify that only some input fields
are retained in the output.
The "commandLine" field in the source-file field list contains the
.I variable.comment
string from the input header.
The field lists of the VIRTUAL fields "source_0",
"source_1", .\ .\ . within "source_0"
(i.e. the subfields "source_0.source_0", "source_0.source_1", etc.)
consist of the embedded source headers of the input file,
converted from FEA format to Esignal.
.PP
If the input file contains the waves generic header items
.I record_freq
and
.I start_time,
the program creates corresponding waves GLOBAL fields
.I recordFreq
and
.I startTime
in the output file.
These are the header items that
.IR xwaves (1\-ESPS)
uses for time alignment of displays of files
with uniformly spaced records.
The value of
.I recordFreq
is just that of
.I record_freq.
The value of
.I startTime
is obtained from that of
.I start_time
by adding an offset that is the time difference
between the first record of the input file
and the first record that is copied to the output.
The offset is 0 unless the
.B \-r
option is used to specify a starting point
other than the first input record.
(See the ``Options'' section.)
In that case, the offset is the result of
subtracting 1 from the first record number
and then dividing by the value of
.I recordFreq.
Like other input generic header items,
.I record_freq
and
.I start_time,
if present, are represented in the output by GLOBAL fields
with the same names.
In this case, the field
.I start_time
is simply a copy of the input generic header item
.I start_time;
it is the GLOBAL field
.I startTime
that contains the correct starting time of the output file.
.SH OPTIONS
.PP
The following options are supported.
Default values are shown in brackets.
.TP
.BI \-a " format \fR[NATIVE]\fP"
specifies whether the output Esignal format is
A\s-3SCII\s+3 (ASCII), native binary (NATIVE\*-the default),
or portable binary (EDR1 or EDR2).
.TP
.BI \-f " field"
specifies the name of an input field to be retained in the output file.
The option may be repeated to specify several input fields to be retained.
If the option is not specified, all input fields are retained.
.TP
.BI \-r " start" : "last \fR[1:(last in file)]\fP"
.TP
.BI \-r " start" :+ "incr"
.TP
.BI \-r " start"
In the first form, a pair of unsigned integers specifies the range
of input records to be processed.
Input record numbers follow ESPS FEA file conventions:
the first record in the file is number 1.
Either
.I start
or
.I last
may be omitted; then the default value is used.
If
.IR last " = " start " + " incr,
the second form (with the plus sign) specifies the same range as the first.
The third form (omitting the colon) specifies a single record.
.TP
.BI \-x " debug_level \fR[0]\fP"
If 
.I debug_level
is positive,
.I fea2esig
writes debugging messages on the standard error output.
Larger values result in more output.
The default is 0, for no output.
.TP
.BI \-A " width"
specifies that ASCII output contains bracketed annotations
that indicate record numbers, field names, and array indices.
The argument indicates the desired approximate width of the output
(number of characters per line).
In the present implementation the approximation is quite rough:
many lines, especially in headers, may well exceed the specified width,
and others, particularly in CHAR data, may be narrower than necessary.
The option is ignored for binary output.
.TP
.BI \-F
specifies output in
.I field order.
If the option is not specified, the output is in
.I type order,
the default.
.SH ESPS PARAMETERS
.PP
The ESPS parameter file is not read.
.SH ESPS COMMON
.PP
The ESPS common file is not accessed.
.SH ESPS HEADERS
.PP
As mentioned under ``Description'' above,
GLOBAL fields corresponding to generic header items in the input
are added to the output file,
and embedded FEA-file source headers, converted to Esignal field lists,
are added as VIRTUAL fields.
The comment string in the input header becomes a GLOBAL field
.I commandLine
in the output.
The input FEA subtype code, if any. is recorded as a GLOBAL field
.I FeaSubtype
in the output.
If generic header items
.I record_freq
and
.I start_time
are present in the input header,
GLOBAL fields
.I recordFreq
and
.I startTime
are added to the output;
see ``Description'' above for details.
.SH FUTURE CHANGES
.PP
Fix some of the problems listed under "Bugs".
.SH EXAMPLES
.PP
.SH ERRORS AND DIAGNOSTICS
.PP
The program exits with a "Usage" message
if an unrecognized option is used on the command line,
or if too many or too few file names are specified.
It may exit with one of the following error messages
if the described conditions occur:
.nf
.RS
can't start before beginning of file;
empty range of records specified;
can't allocate memory for input record;
failure converting input header to field list;
output order neither TYPE_ORDER nor FIELD_ORDER;
can't set field ordering of output;
write header failed.
.RE
.fi
.SH BUGS
.PP
History information is lost if any of the embedded source headers is a
non-FEA ESPS file (e.g. an old-style SD, ANA, or SPEC file,
as distinct from a FEA_SD, FEA_ANA, or FEA_SPEC file).
Confusion is possible if the input file contains fields with names that
.I fea2esig
uses for special purposes\*-for example,
if there is an input field named "commandLine" or "source_0",
or if a tagged input file has a field named "Tag".
.IR A -law
and
.if n .IR mu -law
.if t .IR \(*m -law
encoded FEA_SD files are not recognized as such
and given the appropriate special handling.
.SH REFERENCES
.PP
[1] R. W. Johnson,
``The Esignal File Format'',
Entropic Research Laboratory, Inc., 1995
.SH "SEE ALSO"
.IR esig2fea (1\-ESPS)
.IR Esignal (3\-Esignal)
.IR ESPS (5\-ESPS),
.br
.IR FEA (5\-ESPS),
.PP
.SH AUTHOR
.PP
Rod Johnson


ESPS/general/src/fea2esig/fea2esig.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1995  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  Rod Johnson
 * Checked by:
 * Revised by:
 *
 * Write FEA file data to Esignal file.
 *
 */

static char *sccs_id = "@(#)fea2esig.c	1.4	10/4/97	ERL";

#include <esignal.h>
/* NAME CONFLICT:  THe Esignal data-type codes named
 * DOUBLE, FLOAT, LONG, SHORT, and CHAR
 * are different from the ESPS data-type codes with the same names,
 * and the Esignal type code BOOL has the same name as an ESPS macro.
 * While the Esignal definitions are in effect, declare variables so
 * that the Esignal codes can be referred to as
 * EDOUBLE, EFLOAT, ELONG, ESHORT, ECHAR, and EBOOL
 * instead.  After the inclusion of <esps/esps.h> below,
 * the ESPS definitions of 
 * DOUBLE, FLOAT, LONG, SHORT, CHAR, and BOOL
 * take effect.
 */
const static int EDOUBLE = DOUBLE;
#undef	DOUBLE
const static int EFLOAT = FLOAT;
#undef	FLOAT
const static int ELONG = LONG;
#undef	LONG
const static int ESHORT = SHORT;
#undef	SHORT
const static int ECHAR = CHAR;
#undef	CHAR
const static int EBOOL = BOOL;
#undef BOOL
#include <esps/esps.h>
#include <esps/fea.h>

#define ESPS_TAG "Tag"

#define REQUIRE(test, text) {if (!(test)) {(void) fprintf(stderr, \
"%s: %s - exiting\n", ProgName, text); exit(1);}}

#define SYNTAX \
USAGE("fea2esig [-a format] [-f field]... [-r range] [-x debug_level]\n\t[-A width] [-F] input.fea output.esig");

void	lrange_switch();
char	*savestring();
void	skiprec();
void	fea_skiprec();
int	size_rec();


static FieldList    fea_to_FieldList(struct header *hd, struct fea_data *rec,
				     char **fnames, int copy_sources);
static int	    FindStr(char *str, char **arr);
static int	    EspsTypeToElib(int spstype);
static void	    StrArrToRect(char **strarr,
				 long **dimenp, void **datap);
static FieldSpec    *AddGlobalField(FieldList *list, char *name,
				    int rank, long *dim, int type, void *ptr);
static FieldSpec    *AddSource(FieldList *list,
			      int srcnum, char *name, FieldList source);
static FieldSpec    *AddCommandLine(FieldList *list, char *line);
static char	    *GetCommandLine(int argc, char **argv);


int	    debug_level = 0;
char	    *ProgName = "fea2esig";

main(int   argc,
     char  **argv)
{
    extern int	    optind;	/* for use of getopt() */
    extern char	    *optarg;	/* for use of getopt() */
    int		    ch;		/* command-line option letter */

    int		    outarch = NATIVE;

    char	    **field_names = NULL;
    int		    num_fields = 0;
    int		    alloc_fields = 0;

    int		    rflag = NO;	/* -r option specified? */
    char	    *rrange;	/* arguments of -r option */
    long	    start_rec;	/* starting record number */
    long	    end_rec;	/* ending record number */
    long	    num_recs;	/* number of records to read
				   (0 means all up to end of file) */
    long	    num_read;	/* number of records actually read */

    int		    Aflag = NO;	/* annotate Ascii output? */
    Annot	    *annotate = NULL;
    int		    annwidth = 70;

    char	    *iname;	/* input file name */
    FILE	    *ifile;	/* input stream */
    struct header   *ihd;	/* input file header */
    struct fea_data *irec;	/* input record */

    char	    *oname;	/* output file name */
    FILE	    *ofile;	/* output stream */
    FieldList	    list;	/* output field list */
    int		    outord = TYPE_ORDER;
    FieldSpec	    **ofields;	/* output fields in field or type order */

    double	    rec_freq;
    double	    start_time_offset;
    char	    *fdata;
    double	    *edata;
    int		    type;
    int		    len, i;
    long	    dim[1];
    FieldSpec	    *field;

    FieldList	    source;	/* field list of source file */

    while ((ch = getopt(argc, argv, "a:f:r:x:A:F")) != EOF)
	switch (ch)
	{
	case 'a':
	    outarch = ((!strcmp(optarg, "EDR1")) ? EDR1
		       : (!strcmp(optarg, "EDR2")) ? EDR2
		       : (!strcmp(optarg, "NATIVE")) ? NATIVE
		       : (!strcmp(optarg, "ASCII")) ? ASCII
		       : UNKNOWN);
	    break;
	case 'f':
	    if (num_fields >= alloc_fields)
	    {
		size_t	size;

		alloc_fields = num_fields + 1 + num_fields/2;
		size = (alloc_fields + 1) * sizeof(char *);
		field_names = (char **)
		    ((field_names == NULL)
		     ? malloc(size)
		     : realloc(field_names, size));
	    }
	    field_names[num_fields++] = optarg;
	    field_names[num_fields] = NULL;
	    break;
	case 'r':
	    rflag = YES;
	    rrange = optarg;
	    break;
	case 'x':
	    debug_level = atoi(optarg);
	    break;
	case 'A':
	    Aflag = YES;
	    annwidth = atoi(optarg);
	    break;
	case 'F':
	    outord = FIELD_ORDER;
	    break;
	default:
	    SYNTAX;
	    break;
	}

    if (argc - optind > 2)
    {
	fprintf(stderr,
		"%s: too many file names specified.\n", ProgName);
	SYNTAX;
    }
    if (argc - optind < 2)
    {
	fprintf(stderr,
		"%s: too few file names specified.\n", ProgName);
	SYNTAX;
    }

    iname = eopen(ProgName,
		  argv[optind++], "r", FT_FEA, NONE, &ihd, &ifile);

    oname = argv[optind++];

    start_rec = 1;
    end_rec = LONG_MAX;
    num_recs = 0;

    if (rflag)
    {
	lrange_switch(rrange, &start_rec, &end_rec, 0);
	if (end_rec != LONG_MAX)
	    num_recs = end_rec - start_rec + 1;
    }

    REQUIRE(start_rec >= 1, "can't start before beginning of file");
    REQUIRE(end_rec >= start_rec, "empty range of records specified");

    if (debug_level)
	fprintf(stderr,
		"start_rec: %ld.  end_rec: %ld.  num_recs: %ld.\n",
		start_rec, end_rec, num_recs);

    DebugMsgLevel = debug_level;
    DebugMsgFunc = DebugPrint;

    irec = allo_fea_rec(ihd);

    REQUIRE(irec != NULL, "can't allocate memory for input record");

    list = fea_to_FieldList(ihd, irec, field_names, FALSE);

    REQUIRE(list != NULL,
	    "failure converting input header to field list");

    switch (outord)
    {
    case TYPE_ORDER:
	if (debug_level)
	    fprintf(stderr, "making type-ordered field array.\n");
	ofields = TypeOrder(list);
	break;
    case FIELD_ORDER:
	if (debug_level)
	    fprintf(stderr, "making field-ordered field array.\n");
	ofields = FieldOrder(list);
	break;
    default:
	REQUIRE(0, "output order neither TYPE_ORDER nor FIELD_ORDER");
	break;
    }

    if (debug_level)
	fprintf(stderr, "setting field ordering.\n");

    REQUIRE(SetFieldOrdering(&list, outord),
	    "can't set field ordering of output");

    rec_freq = get_genhd_val("record_freq", ihd, 1.0);

    type = genhd_type("record_freq", &len, ihd);

    if (type != HD_UNDEF)
    {
	field = AddGlobalField(&list, "recordFreq", 0, NULL, EDOUBLE, NULL);
	*(double *) field->data = rec_freq;
    }

    type = genhd_type("start_time", &len, ihd);

    if (type != HD_UNDEF && rec_freq != 0)
    {
	start_time_offset = (start_rec - 1) / rec_freq;

	fdata = (char *) get_genhd("start_time", ihd);
	edata = (double *) type_convert((long) len, fdata, type,
					(char *) NULL, DOUBLE,
					(void (*)()) NULL);

	if (start_time_offset != 0)
	{
	    for (i = 0; i < len; i++)
		edata[i] += start_time_offset;
	}

	dim[0] = len;
	AddGlobalField(&list, "startTime", 1, dim, EDOUBLE, edata);
    }

    source = fea_to_FieldList(ihd, NULL, NULL, TRUE);
    (void) AddSource(&list, 0, iname, source);

    (void) AddCommandLine(&list, GetCommandLine(argc, argv));

    if (debug_level)
	fprintf(stderr, "annwidth %d.\n", annwidth);

    if (Aflag)
    {
	annotate = (Annot *) malloc(sizeof(Annot));
	annotate->position = 0;
	annotate->indent = 0;
	annotate->width = annwidth;
	annotate->recnum = 0;
    }

    if (debug_level)
	fprintf(stderr, "writing Esignal header.\n");

    REQUIRE(OpenOut(oname, list, outarch, &ofile, annotate),
	    "write header failed");
    if (ofile == stdout)
	oname = "<stdout>";

    num_read = start_rec - 1;

    if (debug_level)
	fprintf(stderr, "skipping %ld records.\n", num_read);

/*    skiprec(ifile, num_read, size_rec(ihd)); */
    fea_skiprec(ifile, num_read, ihd);

    while (num_read++ < end_rec && get_fea_rec(irec, ihd, ifile) != EOF)
    {
	if (debug_level > 2)
	    fprintf(stderr, "Record number %ld read.\n", num_read);

	WriteRecord(ofields, outarch, ofile, annotate);
    }

    if (--num_read < end_rec && num_recs != 0)
	fprintf(stderr, "fea2esig: only %ld records read.\n",
		num_read - (start_rec - 1));

    exit(0);
    /*NOTREACHED*/
}


static FieldList
fea_to_FieldList(struct header	    *hdr,
		 struct fea_data    *rec,
		 char		    **field_names,
		 int		    copy_sources)
{
    struct fea_header	*fea;
    FieldList		list;
    int			i, j;
    FieldSpec		*field, *subfield;
    char		*name;
    int			spstype;
    long		*dim;
    int			rank;
    int			type;
    char		**codes;
    char		**gnames;
    int			gnum;
    struct varsize	*var;
    int			nsrc;
    FieldList		source;
    char		*line;
    long		len;


    if (hdr == NULL)
	return NULL;

    if (hdr->common.type != FT_FEA)
	return NULL;

    fea = hdr->hd.fea;

    list = NULL;

    if (hdr->common.tag && FindStr(ESPS_TAG, field_names))
    {
	field = NewFieldSpec(ELONG, 0);
	field->name = savestring(ESPS_TAG);
	field->occurrence = REQUIRED;
	if (rec != NULL)
	    field->data = &rec->tag;
	AddField(&list, field);
    }

    if (debug_level)
	fprintf(stderr, "fea_to_FieldList: field count %d.\n",
		fea->field_count);

    for (i = 0; i < (int) fea->field_count; i++)
    {
	long	size;

	name = fea->names[i];

	if (FindStr(name, field_names))
	{
	    spstype = fea->types[i];
	    size = fea->sizes[i];
	    rank = fea->ranks[i];
	    dim = fea->dimens[i];

	    if (debug_level >= 2)
		fprintf(stderr,
			"fea_to_FieldList: field[%d]: \"%s\".\n",
			i, name);

	    type = EspsTypeToElib(spstype);
	    field = NewFieldSpec(type, rank);
	    if (rank == 1)
		field->dim[0] = size;
	    else
		for (j = 0; j < rank; j++)
		    field->dim[j] = dim[j];
	    field->name = savestring(name);
	    field->occurrence = REQUIRED;
	    if (spstype == CODED)
	    {
		codes = fea->enums[i];
		subfield = NewFieldSpec(ECHAR, 2);
		subfield->name = savestring("enumStrings");
		subfield->occurrence = GLOBAL;
		StrArrToRect(codes, &subfield->dim, &subfield->data);
		AddSubfield(field, subfield);
	    }
	    if (rec != NULL)
		field->data = get_fea_ptr(rec, name, hdr);
	    AddField(&list, field);
	}
    }

    gnames = genhd_list(&gnum, hdr);

    for (i = 0; i < gnum; i++)
    {
	int	size;

	name = gnames[i];
	spstype = genhd_type(name, &size, hdr);

	if (debug_level >= 2)
	    fprintf(stderr,
		    "fea_to_FieldList: global field[%d]: \"%s\".\n",
		    i, name);

	type = EspsTypeToElib(spstype);
	field = NewFieldSpec(type, 1);
	field->dim[0] = size;
	field->name = savestring(name);
	field->occurrence = GLOBAL;
	if (spstype == CODED)
	{
	    codes = genhd_codes(name, hdr);
	    subfield = NewFieldSpec(ECHAR, 2);
	    subfield->name = savestring("enumStrings");
	    subfield->occurrence = GLOBAL;
	    StrArrToRect(codes, &subfield->dim, &subfield->data);
	    AddSubfield(field, subfield);
	}
	field->data = get_genhd(name, hdr);
	AddField(&list, field);
    }

    if (fea->fea_type != NONE)
    {
	field = AddGlobalField(&list, "FeaSubtype",
			       0, NULL, ESHORT, NULL);
	*(short *) field->data = fea->fea_type;
	subfield = NewFieldSpec(ECHAR, 2);
	subfield->name = savestring("enumStrings");
	subfield->occurrence = GLOBAL;
	StrArrToRect(fea_file_type, &subfield->dim, &subfield->data);
	AddSubfield(field, subfield);
    }

    var = &hdr->variable;

    if (copy_sources)
    {
	nsrc = MAX(var->nnames, var->nheads);
	for (i = 0; i < nsrc; i++)
	{
	    source = (i >= var->nheads) ? NULL
		: fea_to_FieldList(var->srchead[i], NULL, NULL, TRUE);
	    name = (i >= var->nnames) ? NULL
		: var->source[i];
	    (void) AddSource(&list, i, name, source);
	}
    }

    if (var->comment != NULL)
    {
	line = savestring(var->comment);
	len = strlen(line);
	if (line[len-1] == '\n')
	    line[len-1] = '\0';
	(void) AddCommandLine(&list, line);
    }

    return list;
}


static int
FindStr(char	*str,
	char	**arr)
{
    int     i;

    if (str == NULL)
	return FALSE;

    if (arr == NULL)
	return TRUE;		/* NULL array implies all fields */

    for (i = 0; arr[i] != NULL; i++)
	if (strcmp(str, arr[i]) == 0)
	    return TRUE;

    return FALSE;
}


/*!*//* NAME CONFLICT BETWEEN ESPS AND ESIGNAL PUBLIC DATA TYPES */

static int
EspsTypeToElib(int spstype)
{
    switch (spstype)
    {
    case DOUBLE:
	return EDOUBLE;
    case FLOAT:
	return EFLOAT;
    case LONG:
	return ELONG;
    case SHORT:
    case CODED:
	return ESHORT;
    case BYTE:
	return SCHAR;
    case CHAR:
    case AFILE:
    case EFILE:
	return ECHAR;
    case DOUBLE_CPLX:
	return DOUBLE_COMPLEX;
    case FLOAT_CPLX:
	return FLOAT_COMPLEX;
    case LONG_CPLX:
	return LONG_COMPLEX;
    case SHORT_CPLX:
	return SHORT_COMPLEX;
    case BYTE_CPLX:
	return SCHAR_COMPLEX;
    default:
	return NO_TYPE;
    }
}


static void
StrArrToRect(char  **strarr,
	     long  **dimenp,
	     void  **datap)
{
    long	len, slen, wid, *dim, siz;
    void	*data;
    long	i, j;
    char	*str, *row;

    wid = 0;
    if (strarr == NULL)
	len = 0;
    else
	for (len = 0; strarr[len] != NULL; len++)
	{
	    if ((slen = strlen(strarr[len])) > wid)
		wid = slen;
	}
    wid += 1;

    dim = (long *) malloc(2*sizeof(long));
    dim[0] = len;
    dim[1] = wid;

    siz = len * wid;
    data = malloc(siz*sizeof(char));
    row = data;
    for (i = 0; i < len; i++)
    {
	str = strarr[i];
	for (j = 0; str[j] != '\0'; j++)
	    row[j] = str[j];
	for ( ; j < wid; j++)
	    row[j] = '\0';
	row += wid;
    }

    if (dimenp)
	*dimenp = dim;
    if (datap)
	*datap = data;
}


/* MOVE TO ESIGNAL PUBLIC I/O LIBRARY */


static FieldSpec *
AddGlobalField(FieldList    *list,
	       char	    *name,
	       int	    rank,
	       long	    *dim,
	       int	    type,
	       void	    *ptr)
{
    FieldSpec	*field;
    int		i;
    long	len;
    size_t	size;
    

    if (name == NULL)
	return NULL;

    if (rank > 0 && dim == NULL)
    {
	DebugMsg(1, "AddGlobalField: NULL dimensions but rank > 0.");
	return NULL;
    }

    field = FindField(*list, name);

    if (field == NULL)
    {
	field = NewFieldSpec(type, rank);

	if (field == NULL)
	{
	    DebugMsg(1, "AddGlobalField: Couldn't create field spec.");
	    return NULL;
	}

	field->name = savestring(name);
	field->occurrence = GLOBAL;

	if (!AddField(list, field))
	{
	    DebugMsg(1, "AddGlobalField: Couldn't add field spec.");
	    return NULL;
	}
    }
    else			/* field != NULL */
    {
	if (field->occurrence != GLOBAL)
	{
	    DebugMsg(1, "AddGlobalField: non-GLOBAL field with same name.");
	    return NULL;
	}

	field->type = type;

	if (field->rank != rank)
	{
	    if (rank == 0)
	    {
		free(field->dim);
		field->dim = NULL;
	    }
	    else
	    {
		field->dim =
		    (long *) ((field->dim == NULL)
			      ? malloc(rank * sizeof(long))
			      : realloc(field->dim, rank * sizeof(long)));

		if (field->dim == NULL)
		{
		    DebugMsg(1,
			     "AddGlobalField: couldn't allocate dimensions.");
		    return NULL;
		}
	    }

	    field->rank = rank;
	}
    }

    len = 1;
    for (i = 0; i < rank; i++)
    {
	field->dim[i] = dim[i];
	len *= dim[i];
    }

    size = len * InternTypeSize(type);

    if (size == 0)
	field->data = NULL;
    else if (ptr != NULL)
	field->data = ptr;
    else
    {
	field->data = malloc(size);

	if (field->data == NULL)
	{
	    DebugMsg(1, "AddGlobalField: couldn't allocate data storage.");
	    return NULL;
	}
    }

    return field;
}


static FieldSpec *
AddSource(FieldList *list,
	  int	    srcnum,
	  char	    *name,
	  FieldList source)
{
    FieldSpec	*field;
    char	srcname[30];
    int		type;
    int		rank;

    if (name == NULL && source == NULL)
	return NULL;

    sprintf(srcname, "source_%d", srcnum);

    if (name == NULL)
    {
	type = NO_TYPE;
	rank = 0;
    }
    else
    {
	type = ECHAR;
	rank = 1;
    }

    field = FindField(*list, srcname);

    if (field == NULL)
    {
	field = NewFieldSpec(type, rank);
	if (field == NULL)
	{
	    DebugMsg(1, "AddSource: Couldn't create field spec.");
	    return NULL;
	}

	field->name = savestring(srcname);
	field->occurrence = VIRTUAL;

	if (!AddField(list, field))
	{
	    DebugMsg(1, "AddSource: Couldn't add field spec.");
	    return NULL;
	}
    }
    else			/* field != NULL */
    {
	if (field->occurrence != VIRTUAL)
	{
	    DebugMsg(1, "AddSource: non-VIRTUAL source_<n> field.");
	    return NULL;
	}

	field->type = type;
	field->rank = rank;

	if (field->dim != NULL)
	{
	    free(field->dim);
	    field->dim = NULL;
	}

	if (name != NULL)
	    field->dim = (long *) malloc(sizeof(long));
    }

    field->subfields = source;

    if (name != NULL)
    {
	field->dim[0] = strlen(name);
	field->data = name;
    }

    return field;
}


static FieldSpec *
AddCommandLine(FieldList    *list,
	       char	    *line)
{
    FieldSpec	*field;


    if (line == NULL)
	return NULL;

    field = FindField(*list, "commandLine");

    if (field == NULL)
    {
	field = NewFieldSpec(ECHAR, 1);
	if (field == NULL)
	{
	    DebugMsg(1, "AddCommandLine: Couldn't create field spec.");
	    return NULL;
	}

	field->dim[0] = strlen(line);
	field->name = savestring("commandLine");
	field->occurrence = GLOBAL;

	if (!AddField(list, field))
	{
	    DebugMsg(1, "AddCommandLine: Couldn't add field spec.");
	    return NULL;
	}
    }
    else			/* field != NULL */
    {
	if (field->occurrence != GLOBAL)
	{
	    DebugMsg(1, "AddCommandLine: non-GLOBAL field \'commandLine\".");
	    return NULL;
	}

	field->type = ECHAR;
	field->rank = 1;
	field->dim = (long *) ((field->dim == NULL)
				? malloc(sizeof(long))
				: realloc(field->dim, sizeof(long)));

	if (field->dim == NULL)
	{
	    DebugMsg(1, "AddCommandLine: couldn't (re)allocate dimension.");
	    return NULL;
	}

	field->dim[0] = 1 + strlen(line);
    }

    field->data = line;

    return field;
}


static char *
GetCommandLine(int  argc,
	       char **argv)
{
    long    len;
    int	    i;
    char    *line;
    char    *arg;


    len = 0;
    for (i = 0; i < argc; i++)
	if (argv[i] != NULL)
	    len += strlen(argv[i]);

    len += 3*argc;		/* Allow for quotes around each item,
				 * blanks between, and null at end. */

    line = (char *) malloc(len * sizeof(char));

    if (line == NULL)
    {
	DebugMsg(1, "GetCommandLine: couldn't allocate string.");
	return NULL;
    }

    line[0] = '\0';

    for (i = 0; i < argc; i++)
	if ((arg = argv[i]) != NULL)
	{
	    if (strpbrk(arg, " \t\n?*$~[];&()|^<>\\\'\"") == NULL)
	    {
		(void) strcat(line, arg);
		(void) strcat(line, " ");
	    }
	    else
	    {
		/* Quote args that contain whitespace or shell special
		 * characters.
		 * (This doesn't handle arguments correctly if they contain
		 * right single quotes (') or, under some shells, newlines).
		 */

		(void) strcat(line, "\'");
		(void) strcat(line, arg);
		(void) strcat(line, "\' ");
	    }
	}

    if (i > 0)
	line[strlen(line) - 1] = '\0'; /* Kill final blank. */

    return line;
}
